tls/s2n_client_hello.c

```diff
@@ -187,15 +177,141 @@ int s2n_collect_client_hello(struct s2n_connection *conn, struct s2n_stuffer *so
 
     struct s2n_client_hello *ch = &conn->client_hello;
 
-    POSIX_GUARD(s2n_stuffer_resize(&ch->raw_message, size));
-    POSIX_GUARD(s2n_stuffer_copy(source, &ch->raw_message, size));
+    POSIX_GUARD(s2n_realloc(&ch->raw_message, size));
+    POSIX_GUARD(s2n_stuffer_read(source, &ch->raw_message));
 
     return 0;
 }
 
+
+static S2N_RESULT s2n_client_hello_verify_for_retry(struct s2n_connection *conn,
+        struct s2n_client_hello *old_ch, struct s2n_client_hello *new_ch,
+        uint8_t *previous_client_random)
+{
+    if (!s2n_is_hello_retry_handshake(conn)) {
+        return S2N_RESULT_OK;
+    }
+
+    /*
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.1.2
+     *# The client will also send a
+     *# ClientHello when the server has responded to its ClientHello with a
+     *# HelloRetryRequest.  In that case, the client MUST send the same
+     *# ClientHello without modification, except as follows:
+     *
+     * All of the exceptions that follow are extensions.
+     * Ignoring the extensions, the client hellos should match /exactly/.
+     */
+    ssize_t old_msg_len = old_ch->raw_message.size;
+    /* Also consider the 2-byte size of the extension list */
+    ssize_t old_extensions_len = old_ch->extensions.raw.size + sizeof(uint16_t);
+    RESULT_ENSURE_GT(old_msg_len, old_extensions_len);
+    size_t verify_len = old_msg_len - old_extensions_len;
+    RESULT_ENSURE(s2n_constant_time_equals(
+        old_ch->raw_message.data,
+        new_ch->raw_message.data,
+        verify_len
```

Comment:
do we need to make sure `new_ch->raw_message.size >= verify_len`?

Reply:
We definitely do D: